src/Processors/QueryPlan/PartsSplitter.cpp

```diff
@@ -49,81 +65,451 @@ class IndexAccess
+                return false;
+
+        return range == other.range && part_index == other.part_index && event == other.event;
+    }
+
+    [[maybe_unused]] bool operator>(const PartsRangesIterator & other) const
+    {
+        if (operator<(other) || operator==(other))
+            return false;
+
+        return true;
+    }
+
+    Values value;
+    MarkRange range;
+    size_t part_index;
+    EventType event;
+};
+
+struct SplitPartsRangesResult
+{
+    RangesInDataParts non_intersecting_parts_ranges;
+    RangesInDataParts intersecting_parts_ranges;
+};
+
+SplitPartsRangesResult splitPartsRanges(RangesInDataParts ranges_in_data_parts)
+{
+    /** Split ranges in data parts into intersecting ranges in data parts and non intersecting ranges in data parts.
+      *
+      * For each marks range we will create 2 events (RangeStart, RangeEnd), add these events into array and sort them by primary key index
+      * value at this event.
+      *
+      * After that we will scan sorted events and maintain current intersecting parts ranges.
+      * If current intersecting parts ranges is 1, for each event (RangeStart, RangeEnd) we can extract non intersecting range
+      * from single part range.
+      *
+      * There can be 4 possible cases:
+      *
+      * 1. RangeStart after RangeStart:
+      *
+      * Example:
+      *
+      * range 1 [----            ...
+      * range 2      [(value_1)    ...
+      *
+      * In this scenario we can extract non intersecting part of range 1. This non intersecting part will have start
+      * of range 1 and end with rightmost mark from range 1 that contains value less than value_1.
+      *
+      * 2. RangeStart after RangeEnd:
+      *
+      * Example:
+      *
+      * range 1   [              ----              ...
+      * range 2   [   (value_1)]
+      * range 3                      [(value_2)    ...
+      *
+      * In this case we can extract non intersecting part of range 1. This non intersecting part will have start
+      * of leftmost mark from range 1 that contains value greater than value_1 and end with rightmost mark from range 1
+      * that contains value less than value_2.
+      *
+      * 3. RangeEnd after RangeStart:
+      *
+      * Example:
+      *
+      * range 1   [----]
+      *
+      * In this case we can extract range 1 as non intersecting.
+      *
+      * 4. RangeEnd after RangeEnd
+      *
+      * Example:
+      *
+      * range 1    [    ...              ----]
+      * range 2    [    ...    (value_1)]
+      *
+      * In this case we can extract non intersecting part of range 1. This non intersecting part will have start
+      * of leftmost mark from range 1 that contains value greater than value_1 and end with range 1 end.
+      *
+      * Additional details:
+      *
+      * 1. If part level is 0, we must process all ranges from this part, because they can contain duplicate primary keys.
```

Comment:
This assumption is incorrect for `ReplacingMergeTree` with `is_deleted` column, and that's why `02814_ReplacingMergeTree_fix_select_final_on_single_partition` was broken
See also https://github.com/ClickHouse/ClickHouse/commit/95e4b0002f2f5c64581268c341ee895c426e1a47